File: rust/src/models/audit_result.rs
use serde::{Deserialize, Serialize};
use crate::models::Origin;

/// A single problem found by a detector. `message` is written for humans and
/// must never contain a variable value.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Finding {
    /// Stable id, e.g. `missing.DATABASE_URL` or `type-mismatch.PORT`.
    pub id: String,
    /// Detector id that produced this finding.
    pub rule_id: String,
    pub severity: Severity,
    pub variable: String,
    pub message: String,
    /// Where the variable was seen; rendered as `path:line`.
    pub locations: Vec<Origin>,
}

impl Finding {
    pub fn new(rule_id: &str, severity: Severity, variable: &str, message: String, locations: Vec<Origin>) -> Self {
        Self {
            id: format!("{}.{}", rule_id, variable),
            rule_id: rule_id.to_string(),
            severity,
            variable: variable.to_string(),
            message,
            locations,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Error,
    Warning,
    Info,
}

impl Severity {
    pub fn as_str(&self) -> &'static str {
        match self {
            Severity::Error => "error",
            Severity::Warning => "warning",
            Severity::Info => "info",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuditSummary {
    pub files_scanned: usize,
    pub variables_found: usize,
    pub errors: usize,
    pub warnings: usize,
    pub infos: usize,
    pub total: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuditResult {
    pub findings: Vec<Finding>,
    pub summary: AuditSummary,
    /// 0 = clean, 1 = errors, (2 is reserved for usage/config errors).
    pub exit_code: u8,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExitContext {
    pub findings: Vec<Finding>,
    pub strict: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AuditFailureKind {
    None,
    Errors,
    Warnings,
    UsageError,
}

File: dotnet/src/Envdoctor/Models/AuditResult.cs
namespace Envdoctor.Models;

/// A single problem found by a detector. `message` is written for humans and
/// must never contain a variable value.
public sealed class Finding
{
    public required string Id { get; set; }
    public required string RuleId { get; set; }
    public Severity Severity { get; set; }
    public required string Variable { get; set; }
    public required string Message { get; set; }
    public List<Origin> Locations { get; set; } = new();

    public static Finding New(
        string ruleId,
        Severity severity,
        string variable,
        string message,
        List<Origin> locations) =>
        new()
        {
            Id = $"{ruleId}.{variable}",
            RuleId = ruleId,
            Severity = severity,
            Variable = variable,
            Message = message,
            Locations = locations,
        };

    public Finding Clone() =>
        new()
        {
            Id = Id,
            RuleId = RuleId,
            Severity = Severity,
            Variable = Variable,
            Message = Message,
            Locations = Locations.Select(o => o.Clone()).ToList(),
        };
}

public enum Severity
{
    Error,
    Warning,
    Info,
}

public static class SeverityExtensions
{
    public static string AsStr(this Severity s) => s switch
    {
        Severity.Error => "error",
        Severity.Warning => "warning",
        Severity.Info => "info",
        _ => "info",
    };
}

public sealed class AuditSummary
{
    public int FilesScanned { get; set; }
    public int VariablesFound { get; set; }
    public int Errors { get; set; }
    public int Warnings { get; set; }
    public int Infos { get; set; }
    public int Total { get; set; }

    public AuditSummary Clone() => (AuditSummary)MemberwiseClone();
}

public sealed class AuditResult
{
    public List<Finding> Findings { get; set; } = new();
    public AuditSummary Summary { get; set; } = new();
    /// 0 = clean, 1 = errors, (2 is reserved for usage/config errors).
    public int ExitCode { get; set; }
}

public sealed class ExitContext
{
    public List<Finding> Findings { get; set; } = new();
    public bool Strict { get; set; }
}
